DCHdigi/src/DCHdigi.cpp

```diff
@@ -0,0 +1,664 @@
+    double z1 = -dch_data->Lhalf;                           // m
+
+    TVector3 p1 (x1,y1,z1);
+
+
+    // point 2
+    double x2 = rz0;                                        // m
+    double y2 = stereosign*rz0*kappa*dch_data->Lhalf;       // m
+    double z2 = dch_data->Lhalf;                            // m
+
+    TVector3 p2 (x2,y2,z2);
+
+    // calculate phi rotation of whole twisted tube, ie, rotation at z=0
+    double phi_z0 = Calculate_wire_phi_z0(ilayer,nphi);
+    p1.RotateZ(phi_z0);
+    p2.RotateZ(phi_z0);
+
+    //--- end calculating wire position
+
+    return (p2-p1).Unit();
+
+}
+
+TVector3 DCHdigi::Calculate_wire_z0_point(int ilayer, int nphi) const
+{
+    auto & l = this->dch_data->database.at(ilayer);
+    double rz0 = l.radius_sw_z0;
+    TVector3 p1 (rz0,0,0);
+    double phi_z0 = Calculate_wire_phi_z0(ilayer,nphi);
+    p1.RotateZ(phi_z0);
+    return p1;
+}
+
+// calculate phi rotation of whole twisted tube, ie, rotation at z=0
+double DCHdigi::Calculate_wire_phi_z0(int ilayer, int nphi) const
+{
+    auto & l = this->dch_data->database.at(ilayer);
+    int ncells = l.nwires/2;
+    double phistep = TMath::TwoPi()/ncells;
+    double phi_z0 = (nphi + 0.25*(l.layer%2))*phistep;
+    return phi_z0;
+}
+
+
+///////////////////////////////////////////////////////////////////////////////////////
+///////////////////////  Calculate vector from hit position to wire   /////////////////
+///////////////////////////////////////////////////////////////////////////////////////
+TVector3 DCHdigi::Calculate_hitpos_to_wire_vector(int ilayer, int nphi, const TVector3 & hit_position /*in cm*/) const
+{
+    // Solution distance from a point to a line given here:
+    // https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Vector_formulation
+    TVector3 n = this->Calculate_wire_vector_ez(ilayer, nphi);
+    TVector3 a = this->Calculate_wire_z0_point (ilayer, nphi);
+    // Remember using cm as natural units of DD4hep consistently!
+    // TVector3 p {hit_position.x()*MM_TO_CM,hit_position.y()*MM_TO_CM,hit_position.z()*MM_TO_CM};
+
+    TVector3 a_minus_p = a - hit_position;
+    double a_minus_p_dot_n = a_minus_p.Dot( n );
+    TVector3 scaled_n = a_minus_p_dot_n * n;
+    //hit_to_wire_vector = a_minus_p - scaled_n;
+    return (a_minus_p - scaled_n);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////
+///////////////////////       CalculateNClusters       ////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////
+std::pair<uint32_t,uint32_t> DCHdigi::CalculateClusters(const edm4hep::SimTrackerHit & input_sim_hit) const
+{
+    std::pair<uint32_t,uint32_t> return_values = {0,0};
+    uint32_t & Ncl  = return_values.first;
+    uint32_t & ClSz = return_values.second;
+    //_________________SET NECESSARY PARAMETERS FOR THE CLS ALGORITHM-----WALAA_________________//
```

Comment:
All these parameters will be set at every call of the function, since many are hard coded, would'nt it be better to define them in the header once and for all?